api/artillery-load-test.yml

```diff
@@ -0,0 +1,53 @@
+config:
+  target: http://localhost:8080
+  tls:
+    rejectUnauthorized: false
+  http:
+    timeout: 10
+  phases:
+    - duration: 30
+      arrivalRate: 2
+      maxVusers: 50
+      name: Warm up phase
+    - duration: 30
+      arrivalRate: 28
+      maxVusers: 250
+      name: Ramp up load
+    - duration: 30
+      arrivalRate: 56
+      maxVusers: 1000
+      name: Spike phase
+  environments:
+    prod:
+      target: http://api-prod-342430507.us-east-1.elb.amazonaws.com
+    staging:
+      target: http://api-staging-770293556.us-east-1.elb.amazonaws.com
+    dev:
+      target: http://api-dev-1839587515.us-east-1.elb.amazonaws.com
+  plugins:
+    expect:
+      outputFormat: prettyError
+    ensure: {}
+    apdex: {}
+    metrics-by-endpoint: {}
+  apdex:
+    threshold: 100
+scenarios:
```

Comment:
Shouldn't we have the opportunity endpoints included here? I don't think any of these endpoints are representative of testing we are scaled enough.

Reply:
@chouinar I can work on that, it'll just increase the scope because I need to figure out how to handle API keys

Reply:
found the docs: https://www.artillery.io/docs/reference/test-script#using-environment-varables

Reply:
@chouinar search endpoint load test: https://github.com/HHS/simpler-grants-gov/pull/1372/commits/1064eecf48b5ce30f71306b803805b8ec6880058

Unsurprisingly, we very much fail this load test right now.

Reply:
What do you mean by fail? What were the metrics?

Reply:
@chouinar here are the metrics, as a screenshot because I couldn't get the text to load. It's probably too long.

<img width="596" alt="image" src="https://github.com/HHS/simpler-grants-gov/assets/5768468/83390825-3b13-403a-aff1-5a0bae3b5116">

The scores are pretty mediocre. p99 or 9999.2 for the spike load, for example.

They produce an "Apdex score" of "0.13291139240506328 (unacceptable)", which I am translating to "fail".

Reply:
Which environment did you run it against?

Reply:
The screenshot above is against prod

Reply:
I ran it against every environment, though. The scores are similar.

Reply:
Something is off - was curious so I went to see metrics, and I'm not seeing anything that would make it look like the search endpoint actually got called? Healthcheck was called a lot, but not search (all on log insights over the last 3 hours):

![Screenshot 2024-03-01 at 3 39 45 PM](https://github.com/HHS/simpler-grants-gov/assets/46358556/a7a0b9a0-a0e6-40bd-a755-28cf8c63a94e)

![Screenshot 2024-03-01 at 3 40 23 PM](https://github.com/HHS/simpler-grants-gov/assets/46358556/a1fcf468-1a01-4b32-b291-9bf05cef24bb)

Reply:
Querying against prod, I do see 16k requests, not sure why dev has nothing

Reply:
@chouinar I haven't tested `search` against dev yet. I tested search against prod.

Here's a plaintext snippet of the tool working in debug mode to confirm that I am in-fact testing search in prod:

```
  "url": "http://api-prod-342430507.us-east-1.elb.amazonaws.com/v0/opportunities/search",
  "method": "POST",
  "headers": {
    "user-agent": "Artillery (https://artillery.io)",
    "x-auth": "VxnmtL8aQQhwThxNqfaK"
  },
  "json": {
    "paging": {
      "page_offset": 1,
      "page_size": 25
    },
    "sorting": {
      "order_by": "opportunity_id",
      "sort_direction": "ascending"
    }
  }
}
2024-03-01T20:43:09.777Z http:response {
  "date": "Fri, 01 Mar 2024 20:43:09 GMT",
  "content-type": "application/json",
  "content-length": "10500",
  "connection": "keep-alive",
  "server": "gunicorn"
}
2024-03-01T20:43:09.777Z http:response "{\n  \"data\": [\n    {\n      \"agency\": \"DOI-USGS1\",\n      \"category\": \"D\",\n      \"category_explanation\": null,\n      \"created_at\": null,\n      \"modified_comments\": null,\n      \"opportunity_id\": 15936,\n      \"opportunity_number\": \"07HQPA0028\",\n      \"opportunity_title\": \"Cooperative Ecosystem Studies Unit, Rocky Mountain CESU\",\n      \"revision_number\": 0,\n      \"updated_at\": null\n    },\n    {\n      \"agency\": \"ED\",\n      \"category\": \"D\",\n      \"category_explanation\": null,\n      \"created_at\": null,\n      \"modified_comments\": null,\n      \"opportunity_id\": 15937,\n      \"opportunity_number\": \"ED-GRANTS-070907-001\",\n      \"opportunity_title\": \"Youth violence and related issues in persistently dangerous schools CFDA 84.184V\",\n      \"revision_number\": 0,\n      \"updated_at\": null\n    },\n    {\n      \"agency\": \"ED\",\n      \"category\": \"D\",\n      \"category_explanation\": null,\n      \"created_at\": null,\n      \"modified_comments\": null,\n      \"opportunity_id\": 15939,\n      \"opportunity_number\": \"ED-GRANTS-070907-002\",\n      \"opportunity_title\": \"State data collection; technical assistance CFDA 84.373X\",\n      \"revision_number\": 0,\n      \"updated_at\": null\n    },\n    {\n      \"agency\": \"NASA-HQ\",\n      \"category\": \"D\",\n      \"category_explanation\": null,\n      \"created_at\": null,\n      \"modified_comments\": null,\n      \"opportunity_id\": 15941,\n      \"opportunity_number\": \"NNH07ZEA001N-EDL1\",\n      \"opportunity_title\": \"ROA 2007: A.6 Entry, Descent, and Landing 1\",\n      \"revision_number\": 0,\n      \"updated_at\": null\n    },\n    {\n      \"agency\": \"USDOJ-OJP-NIJ\",\n      \"category\": \"D\",\n      \"category_explanation\": null,\n      \"created_at\": null,\n      \"modified_comments\": null,\n      \"opportunity_id\": 16133,\n      \"opportunity_number\": \"2007-NIJ-1667\",\n      \"opportunity_title\": \"NIJ FY08 Institutional Corrections Technology\",\n      \"revision_number\": 0,\n      \"updated_at\": null\n    },\n    {\n      \"agency\": \"IVV\",\n      \"category\": \"D\",\n      \"category_explanation\": null,\n      \"created_at\": null,\n      \"modified_comments\": null,\n      \"opportunity_id\": 16134,\n      \"opportunity_number\": \"IVV-07122007-PACKAGE1\",\n      \"opportunity_title\": \"This is a test package. Please DO NOT apply!\",\n      \"revision_number\": 0,\n      \"updated_at\": null\n    },\n    {\n      \"agency\": \"CNCS\",\n      \"category\": \"D\",\n      \"category_explanation\": null,\n      \"created_at\": null,\n      \"modified_comments\": \" Changed the closing date to Sunday.           \",\n      \"opportunity_id\": 17675,\n      \"opportunity_number\": \"CNCS-GRANTS-071307-001\",\n      \"opportunity_title\": \"Volunteer Management\",\n      \"revision_number\": 5,\n      \"updated_at\": \"2007-09-07T00:00:00\"\n    },\n    {\n      \"agency\": \"DOT-PHMSA\",\n      \"category\": \"D\",\n      \"category_explanation\": null,\n      \"created_at\": null,\n      \"modified_comments\": null,\n      \"opportunity_id\": 25767,\n      \"opportunity_number\": \"DTPH56-07-SN-0003\",
```

Reply:
FYI search doesn't seems to work on staging right now, probably because the code hasn't been deployed in a while. I think that's fine, we had a ticket for deploying the latest code to staging.

Reply:
Just ran the search test against dev

Reply:
Yeah, I found the metrics. Looks like it is maxing out on CPU. We have a single ECS task with 1 vCPU. We've currently got it hard-code to 2 workers, and 4 threads each (that should be calculated from the CPU, that needs to be fixed). So, at most we'd probably handle 8 requests at a time - maybe more, but that's all we have for threads.

From the looks of this artillery test it scales up to a lot more requests/sec than we'd be able to handle as configured.

Looking at past projects, we usually ran with 2 vCPUs per ECS task - and had autoscaling to handle increased load, although I want to say we had 3 ECS tasks minimum?

Reply:
That's useful context for when we want to scale up for an actual release - but I think that for the moment its reasonable to keep everything scaled down very low to save $$$. We don't really have users yet.

My understanding (and correct me if this is wrong) was that we are setting up load testing _**now**_ because _**in the future**_ we want to use it to actually scale our services.

Reply:
I'm not sure - we should probably know what level of traffic we can handle now / what we might roughly expect and make sure those are aligned. I don't expect that number to be high.

This did at least make me aware of the incorrect configuration on the API itself that needs to be fixed before we try any additional scaling. I'll make a ticket for that.

Reply:
I'll make a ticket for scaling for the v0 release (...)

Reply:
My "actually scale the thing" ticket: https://github.com/HHS/simpler-grants-gov/issues/1390